core/o11y/src/io_tracer.rs

```diff
@@ -0,0 +1,254 @@
+use std::collections::HashMap;
+use std::io::Write;
+use std::{fs::File, sync::Mutex};
+use tracing::{span, Subscriber};
+use tracing_subscriber::{registry::LookupSpan, Layer};
+
+/// Tracing layer that produces a record of IO operations.
+pub struct IoTraceLayer {
+    file: Mutex<File>,
+}
+
+enum IoEventType {
+    StorageOp(StorageOp),
+    DbOp(DbOp),
+}
+#[derive(strum::Display)]
+#[strum(serialize_all = "SCREAMING_SNAKE_CASE")]
+enum StorageOp {
+    Read,
+    Write,
+    Other,
+}
+#[derive(strum::Display)]
+#[strum(serialize_all = "SCREAMING_SNAKE_CASE")]
+enum DbOp {
+    Get,
+    Insert,
+    Set,
+    UpdateRc,
+    Delete,
+    DeleteAll,
+    Other,
+}
+
+/// Formatted but not-yet printed output lines.
+///
+/// Events are bundled together and only printed after the enclosing span exits.
+/// This allows to print information at the tpo that is only available later on.
+///
+/// Note: Type used as key in `AnyMap` inside span extensions.
+struct OutputBuffer(Vec<BufferedLine>);
+
+/// Formatted but not-yet printed output line.
+struct BufferedLine {
+    indent: usize,
+    output_line: String,
+}
+
+/// Information added to a span through events happening within.
+#[derive(Default)]
+struct SpanInfo {
+    key_values: Vec<String>,
+    counts: HashMap<String, u64>,
+}
+
+impl<S: Subscriber + for<'span> LookupSpan<'span>> Layer<S> for IoTraceLayer {
+    fn on_enter(&self, id: &span::Id, ctx: tracing_subscriber::layer::Context<'_, S>) {
+        let span = ctx.span(id).unwrap();
+        span.extensions_mut().replace(OutputBuffer(vec![]));
+        span.extensions_mut().replace(SpanInfo::default());
```

Comment:
1. Ok, I changed that. But for my education, is there a reason why `on_new_span` is preferred over `on_enter`? My thinking was if spans are reused somewhere, clearing every time it is entered should be the more robust solution. But then again, I have very little experience with `tracing` and how it is typically used.
2. I can't seem find `on_register` or something with a similar name anywhere, where is it defined? [Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html) doesn't have it, nor can I find it anywhere else in the docs for the involved tracing crates.

Reply:
> But for my education, is there a reason why on_new_span is preferred over on_enter? My thinking was if spans are reused somewhere, clearing every time it is entered should be the more robust solution. But then again, I have very little experience with tracing and how it is typically used.

For instrumented `Future`s the span will be `enter`ed every single time a future is polled. This is can be an extremely frequent operation, so ideally it should be relatively extremely cheap so that the overhead on async code isn’t outrageous. Outside of that, not dealing with the overhead for spans that are manually entered and exited multiple times is also nice. Sorry for not explaining it in the first comment.

> I can't seem find on_register or something with a similar name anywhere, where is it defined? [Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html) doesn't have it, nor can I find it anywhere else in the docs for the involved tracing crates.

Sorry, my bad. I meant `on_record` here. In practice what I’m saying here is that

```
let span = tracing::info_span!("name", size = tracing::field::Empty).entered();
span.record("size", &124);
```

should produce largely the same results as

```
let span = tracing::info_span!("name", size = &124).entered();
```

Reply:
Thank you for the explanations! This helps a lot.

I added the implementation for records as well.